### PR TITLE
Annotate implementations of IDynamicTypeSymbol

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for review

**Comment by jcouv** [comment]
>         Cci.INamespaceTypeReference Cci.ITypeReference.AsNamespaceTypeReference

Consider annotating `Cci.ITypeReference` in this PR too. I suspect this API should return a nullable, even though this specific implementation never returns `null`

---
Refers to: src/Compilers/CSharp/Portable/Emitter/Model/DynamicTypeSymbolAdapter.cs:59 in 55f0487. [](commit_id = 55f0487bb8c6981c71392841d225c0ccb4701753, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/DynamicTypeSymbol.cs [hunk 1]
```diff
@@ -190,9 +193,9 @@ public override int GetHashCode()
             return (int)Microsoft.CodeAnalysis.SpecialType.System_Object;
         }
 
-        internal override bool Equals(TypeSymbol t2, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool> isValueTypeOverrideOpt = null)
+        internal override bool Equals(TypeSymbol? t2, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool>? isValueTypeOverrideOpt = null)
         {
-            if ((object)t2 == null)
+            if ((object?)t2 == null)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/DynamicTypeSymbol.cs]: >(object?)t2 == null [](start = 16, length = 19)

nit: prefer `is` check

Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/DynamicTypeSymbol.cs]: I tried to minimize extraneous changes in all of these PRs.

**Comment by sharwell** [comment]
> Consider annotating `Cci.ITypeReference` in this PR too. I suspect this API should return a nullable, even though this specific implementation never returns null

The API is annotated and returns a nullable type. This implementation returns a more specific non-nullable type.

**Pull request merged**